[core] adding stl io two missing classes

josep-m-carbonell opened the pull request:
Adds  operator << for std::set and std::weak_ptr

josep-m-carbonell committed:
adding stl io two missing classes

philbucher commented on kratos/utilities/stl_io.h:
```diff
@@ -39,6 +41,34 @@ std::ostream& operator<<(std::ostream& rOStream, const std::vector<T>& rVec) {
 }
 
+template<class T>
+std::ostream& operator <<(std::ostream& rOStream, const std::set<T>& rVec) {
```
```suggestion
std::ostream& operator <<(std::ostream& rOStream, const std::set<T>& rSet) {
```

josep-m-carbonell replied:
Sorry for the failure of the test. I will correct it. I wanted to know if adding this operator here is correct. The code was not compiling with one of this types as variable. Do you know where is shared_ptr defining this operator?

philbucher replied:
I think this is the correct place

Unfortunately I don't know where shared_ptr is defining this operator
@roigcarlo ?

philbucher commented on kratos/utilities/stl_io.h:
```diff
@@ -39,6 +41,34 @@ std::ostream& operator<<(std::ostream& rOStream, const std::vector<T>& rVec) {
+template<class T>
+std::ostream& operator <<(std::ostream& rOStream, const std::set<T>& rVec) {
+
+  std::size_t vector_size = rVec.size();
```
```suggestion
  const std::size_t set_size = rSet.size();
```

philbucher reviewed (commented):
minor comments, otherwise nice addition, including tests, :+1: 

Btw your test fails

philbucher committed:
Update kratos/utilities/stl_io.h

Co-Authored-By: josep-m-carbonell <cpuigbo@cimne.upc.edu>

philbucher committed:
Update kratos/utilities/stl_io.h

Co-Authored-By: josep-m-carbonell <cpuigbo@cimne.upc.edu>

josep-m-carbonell committed:
Changing rVec to rSet

philbucher reviewed (dismissed):
looks good to me :+1:

josep-m-carbonell commented:
I do not have a computer now. I will fix tests on Monday. Do not merge. Thank you.

josep-m-carbonell committed:
fixing stl_io tests for set and weak_ptr

adding also a new STRING_CHECK instead of the standard check to print compared values correctly.

josep-m-carbonell commented:
Ok, I think is working now, I have added an extra macro to check for strings (now the compared values are shown when tests fail). If appveyor and travis builds are successful let's please re-approve the PR. Thank you!

philbucher commented on kratos/includes/checks.h:
```diff
@@ -27,6 +27,9 @@
 #define KRATOS_CHECK_EQUAL(a,b) if(!(a == b)) KRATOS_ERROR << "Check failed because " << #a << " is not equal to " << #b
 #define KRATOS_CHECK_NOT_EQUAL(a,b) if(a == b) KRATOS_ERROR << "Check failed because " << #a << " is equal to " << #b
 
+#define KRATOS_CHECK_STRING_EQUAL(a,b) if(a.compare(b) != 0) KRATOS_ERROR << "Check failed because \"" << a << "\" is not equal to \"" << b << "\""
```
just FYI, compare is the same as == for std::strings

josep-m-carbonell replied:
The problem was the output of KRATOS_ERROR was "#a" instead of "a".
